Fix false positive ADTS probes

timo-schluessler opened the pull request:
### This PR will...
fix a bug: false positive ADTS probe in MP3 stream (may happen in any data stream).

### Why is this Pull Request needed?
To fix the bug.

### Are there any points in the code the reviewer needs to double check?
I removed the check whether a frame ends at the end of the data supplied to probe(). Instead probe() only returns true if there are at least three valid headers and two valid frames in between. I think this makes sense because every HLS chunk should have at least three full frames in it.
The main point of the bug fix is the check against zero sized frames. And to fix the header length check which was broken as I think.

### Resolves issues:

### Checklist

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] API or design changes are documented in API.md

itsjamie commented:
I'm a little confused, where is the ADTS probe running on a MP3 stream?

I agree, the header length check seems off, do you have a test case where the old value causes a problem? I'd love to add it as a unit test for the probe code.

---

The discussion around changing the probe code to look for more data I can understand why you would want to do this if you were running this code in a magic byte style algorithim when checking many possible formats. But, you could achieve that using the single header check as a user without changing the hls.js implementation. But, open to discussion on this one.

itsjamie commented:
Currently these two tests are failing.

```
 it('should return true if it contains the entire ADTS frame', function () {
    const data = new Uint8Array(new ArrayBuffer(16));
    data[0] = 0xff;
    data[1] = 0xf0; // protection_absent = 0
    data[4] = 0x02; // frame_length is 16
    expect(probe(data, 0)).to.be.true;
  });

  it('should return true if it contains an valid following frame header', function () {
    const data = new Uint8Array(new ArrayBuffer(18));
    data[0] = 0xff;
    data[1] = 0xf0; // protection_absent = 0
    data[4] = 0x02; // frame_length is 16
    data[16] = 0xff;
    data[17] = 0xf0;
    expect(probe(data, 0)).to.be.true;
  });
```

timo-schluessler commented:
As far as I understood audio streams are always tested against TS/MP4/AAC/MP3 in this order (see src/demux/demuxer-inline.js:76). That's why my MP3 stream was accidentally recognized as AAC ('ADTS sync word found') because it contains the byte sequence 0xff 0xf9 0x45 0x00 0x00 0x1d at an arbitrary position (extract.mp3 at offset 0x64).
[extract.zip](https://github.com/video-dev/hls.js/files/4460842/extract.zip)

Regarding this header length check I constructed a test case that doesn't contain a full header but still lets probe return true.

Regarding the real bug I added a test case that covers this one. (Sorry for the failing tests, I didn't have CHROME_BIN setup correctly and thus didn't notice these tests.)

Regarding the "if we ever probe not at the beginning of the file" I think this is open to discussion. For my 'error case' it is sufficient to add the test against zero length frames. On the other hand I think there might be other false positives if we only check for one header and the correct two bytes of a header after the frame length.

I updated my branch with another version that only checks for one frame and the following header (as before) but also checks the header length and the frame length against invalid short frames.
